### PR TITLE
Implement batch matrix inverse

### PR DESCRIPTION
**Pull request by karol-arndt**
I was recently working on some Kalman filter stuff and found myself in need of a batch matrix inverse, so I implemented it (doing it in a Python for loop is incredibly slow, especially with CUDA). Since cuBLAS has such functionality already implemented (and the standard inverse function just passes 1 as the batch size), it's just a matter of allocating some buffers and passing data to the appropriate cuBLAS functions. The implementation is based on the `btrifact` (batch LU factorization using `getrf`) function. I also added a CPU implementation (which is really just a for loop) for the sake of completeness. 

I figured that this might be useful for other people, so I'm sharing it here. This is my first contribution to PyTorch and I'm not very experienced with CUDA programming, so all comments regarding the code are most welcome.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
A general `getrf` wrapper may also be useful, e.g. for porting Randomized PCA functions (https://github.com/pytorch/pytorch/issues/8049)...

**Comment by karol-arndt** [comment]
@vadimkantorov isn't that what `btrifact` does? Looking at the code, it only wraps the call to `getrf` with some code to ensure column-major, manage the device buffers and check the error codes.

**Comment by vadimkantorov** [comment]
@karol-arndt You are right! Thanks for the tip :) This makes porting that fbpca code trivial to PyTorch.

**Comment by fmassa** [comment]
Thanks for the PR!
I have a quick suggestion on the python interface: instead of making a new `binverse` function, can't we instead extend support for `inverse` to support arbitrary batch dimensions (where we view all the batch dimensions as 1 and then view back after the result). What do you think?

**Comment by karol-arndt** [comment]
Well, this indeed sounds like a natural and intuitive extension of the current `inverse` method, and wouldn't cause any extra clutter in the Python interface. It's also how the API of _another popular tensor library_ works.
On the other hand, currently most - if not all - functions that operate on batch data have a name prefixed with a b (`bmm`, `btrifact`, `btrisolve`...), and extending `inverse` to work both with batches and single n-by-n matrices would break that convention.
Overall, it's hard to say which is preferable; perhaps some feedback from the code owners could help to solve this problem.

**Comment by fmassa** [comment]
I think the trend will probably be to move away from the `b` prefix (which is in some sense legacy), in favor of functions that handle batches (like `matmul` is the successor of `mm` and `bmm`).

But let's see what the others think about it.

**Comment by zou3519** [comment]
I agree with @fmassa -- I think we're trying to move away from the 'b' prefix to simplify the API, and I would prefer a batch inverse function be built into the inverse function.

**Comment by ezyang** [comment]
How hard would it be to implement this in ATen versus TH? (Asking for information.)

**Comment by karol-arndt** [comment]
@fmassa In that case, I'll adopt the code to extend the `inverse` function.

@ezyang I've never written any ATen code before. I wouldn't expect it to be particularly difficult though, the API seems to be more friendly than TH. If the current trend is to write new code in ATen, I can try to port the implementation in the coming days. Should I use MKL for the CPU code?

**Comment by ezyang** [comment]
Yep, we're trying to do everything in ATen as much as possible. Sometimes it's not possible, but when it is, it's preferred.

If MKL has got a good CPU implementation, it's definitely a good pick.

**Comment by karol-arndt** [comment]
@ezyang I implemented the inverse methods in ATen. I had to add a cuBLAS handle to ATen context (it's shared with the THC handle, similarly to the cuSPARSE handle). 

Worth noting - in addition to cuBLAS version, THC also had a Magma-based implementation of inverse, which I didn't reimplement when porting to ATen. Is it still needed and should it be reimplemented?

@fmassa As you asked, the code now works with arbitrary number of batch dimensions as an extension of the previous `inverse` method.

File: aten/src/ATen/native/cuda/Inverse.cu [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        auto input_data = input_colmajor.data<scalar_t>();
+        auto output_data = result.data<scalar_t>();
+        bool errorOccurred = false;
+        int *info_cpu = new int[batch_size];
+
+        scalar_t **input_gpu;
+        scalar_t **output_gpu;
+        scalar_t **input_ptrs = new scalar_t*[batch_size];
+        scalar_t **output_ptrs = new scalar_t*[batch_size];
+        AT_CUDA_CHECK(cudaMalloc(&input_gpu, batch_size*sizeof(scalar_t*)));
```
Comment (fmassa) [inline aten/src/ATen/native/cuda/Inverse.cu]: could you please replace the instances of `cudaMalloc` with `THCudaMalloc`, and `cudaFree` with `THCudaFree`? @ezyang , I don't think we expose those in ATen yet, right?

**Comment by soumith** [comment]
@karol-arndt the MAGMA based inverse is MUCH faster than the cublas version for many sizes and overall has a better performance profile (as can be attested to by @martinarjovsky who's been reporting this). If it's not too much of an ask, porting the magma bindings would be good as well.

**Comment by karol-arndt** [comment]
Well, this PR definitely needs more work anyway, as some of the tests are currently failing (the results appear to be transposes of the correct ones, which seems like a data alignment issue on some platforms). I currently don't have the time to work on this anymore - and the current implementation is good enough for me to continue my research. I will most likely return to this in a few weeks, but if someone has the time and energy to fix the issue and add the MAGMA version, that would be great ;)

**Comment by weiyangfb** [comment]
I believe there is a continue work of this PR at https://github.com/pytorch/pytorch/pull/9949, Thanks @karol-arndt for implementing this feature and moved it into ATen!

**Comment by vishwakftw** [comment]
I think this can be closed now, since batch inverse is now part of master.

**Comment by fmassa** [comment]
Thanks @karol-arndt for the original implementation!

**Pull request closed without merging**